msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java

```diff
@@ -486,4 +492,34 @@ public static GenerateShrCommandParameters createGenerateShrCommandParameters(
                 .popParameters(popParameters)
                 .build();
     }
+
+    /**
+     * Adds additional query string parameters to original list.
+     * @param queryStringParameters extra query string parameters inputted.
+     * @param configuration configuration created from MSAL config file.
+     * @return combined list of query string parameters.
+     */
+    @Nullable
+    public static List<Map.Entry<String, String>> appendToExtraQueryParametersIfWebAuthnCapable(
+            @Nullable final List<Map.Entry<String, String>> queryStringParameters,
+            @NonNull final PublicClientApplicationConfiguration configuration) {
+        if (configuration.isWebauthnCapable()) {
+            final Map.Entry<String, String> webauthnExtraParameter = new AbstractMap.SimpleEntry<>(
+                    FidoConstants.WEBAUTHN_QUERY_PARAMETER_FIELD,
+                    FidoConstants.WEBAUTHN_QUERY_PARAMETER_VALUE);
+            if (queryStringParameters == null) {
+                return new ArrayList<>(Collections.singletonList(webauthnExtraParameter));
+            }
+            if (!queryStringParameters.contains(webauthnExtraParameter)) {
```

Comment:
>!queryStringParameters.contains(webauthnExtraParameter) [](http://example.com/codeflow?start=16&length=55)

Q: Can input query parameters contain webauthn extra parameter as valid contract to pass the webauthn param value? 
Technically, developer can pass it. If they do, do you want to prefer that over config set (if set)?

Reply:
Yes, it's valid for the developer to pass in the WebAuthn param value into extraQP in AcquireTokenParameters. Though unless they have a good reason to only pass through the parameters (like their own flighting maybe), it'll be recommended to set the property in MSAL config file so that the developer doesn't have to repeatedly add the query parameter on all acquireToken calls.
Like you said, since an app can decide to set both, the above conditional just leaves the List as is in that case.